dataplane/linux/vxlan_mgr.go

```diff
@@ -0,0 +1,379 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package intdataplane
+
+import (
+	"fmt"
+	"net"
+	"reflect"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/vishvananda/netlink"
+
+	"github.com/projectcalico/felix/ip"
+	"github.com/projectcalico/felix/proto"
+	"github.com/projectcalico/felix/routetable"
+)
+
+type vxlanManager struct {
+	sync.Mutex
+
+	// Our dependencies.
+	hostname   string
+	routeTable routeTable
+
+	// Hold pending updates.
+	routesByDest map[string]*proto.RouteUpdate
+	vtepsByNode  map[string]*proto.VXLANTunnelEndpointUpdate
+
+	// Holds this node's VTEP information.
+	myVTEP *proto.VXLANTunnelEndpointUpdate
+
+	// VXLAN configuration.
+	vxlanDevice string
+	vxlanID     int
+	vxlanPort   int
+
+	// Indicates if configuration has changed since the last apply.
+	dirty chan bool
+}
+
+func newVXLANManager(hostname string, routeTable routeTable, deviceName string, vxlanID, port int) *vxlanManager {
+	return &vxlanManager{
+		hostname:     hostname,
+		routeTable:   routeTable,
+		routesByDest: map[string]*proto.RouteUpdate{},
+		vtepsByNode:  map[string]*proto.VXLANTunnelEndpointUpdate{},
+		vxlanDevice:  deviceName,
+		vxlanID:      vxlanID,
+		vxlanPort:    port,
+		dirty:        make(chan bool, 1),
+	}
+}
+
+func (m *vxlanManager) OnUpdate(protoBufMsg interface{}) {
+	switch msg := protoBufMsg.(type) {
+	case *proto.RouteUpdate:
+		if msg.Type == proto.RouteType_VXLAN {
+			logrus.WithField("msg", msg).Debugf("VXLAN data plane received route update")
+			m.routesByDest[msg.Dst] = msg
+			m.setDirty()
+		}
+	case *proto.RouteRemove:
+		if msg.Type == proto.RouteType_VXLAN {
+			logrus.WithField("msg", msg).Debugf("VXLAN data plane received route remove")
+			delete(m.routesByDest, msg.Dst)
+			m.setDirty()
+		}
+	case *proto.VXLANTunnelEndpointUpdate:
+		logrus.WithField("msg", msg).Debug("VXLAN data plane received VTEP update")
+		if msg.Node == m.hostname {
+			m.setLocalVTEP(msg)
+		} else {
+			m.vtepsByNode[msg.Node] = msg
+		}
+		m.setDirty()
+	case *proto.VXLANTunnelEndpointRemove:
+		logrus.WithField("msg", msg).Debug("VXLAN data plane received VTEP remove")
+		if msg.Node == m.hostname {
+			m.setLocalVTEP(nil)
+		} else {
+			delete(m.vtepsByNode, msg.Node)
+		}
+		m.setDirty()
+	}
+}
+
+func (m *vxlanManager) setLocalVTEP(vtep *proto.VXLANTunnelEndpointUpdate) {
+	m.Lock()
+	defer m.Unlock()
+	m.myVTEP = vtep
+}
+
+func (m *vxlanManager) getLocalVTEP() *proto.VXLANTunnelEndpointUpdate {
+	m.Lock()
+	defer m.Unlock()
+	return m.myVTEP
+}
+
+func (m *vxlanManager) setDirty() {
+	select {
+	case m.dirty <- true:
```

Comment:
Why do you need this channel machinery?  I can't see anywhere that it gets used from another goroutine

Reply:
Hrm, for some reason I thought `CompleteDeferredWork` was called from another goroutine... but looking at it now I don't see that it is... I'll clean it up.